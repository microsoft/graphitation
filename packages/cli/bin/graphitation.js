#!/usr/bin/env node

(function () {
  require("@graphitation/cli/lib/graphitation")
    .graphitation()
    .parseAsync(process.argv);
})();
