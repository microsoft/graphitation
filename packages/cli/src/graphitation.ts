import path from "path";
import fs from "fs/promises";
import fsSync from "fs";
import ts from "typescript";
import { program, Command } from "commander";
import { extractImplicitTypesToTypescript } from "@graphitation/supermassive-extractors";
import { parse, print } from "graphql";
import { generateTS } from "@graphitation/ts-codegen";
import {
  mergeSchemas as mergeSchemasImpl,
  printMergeSchemaError,
  FileSystemModuleLoader,
} from "@graphitation/merge-schemas";
import * as glob from "fast-glob";

type GenerateInterfacesOptions = {
  outputDir?: string;
  contextImport?: string;
  contextName?: string;
  enumsImport?: string;
  legacy?: boolean;
  legacyModels?: boolean;
  scope?: string;
};

type MergeSchemasOptions = {
  output?: string;
  ignoreEntryPointMissingTypes?: boolean;
};

const PREPEND_TO_INTERFACES = `/* eslint-disable */ \n// This file was automatically generated (by @graphitation/cli) and should not be edited.\n`;
const PREPEND_TO_GRAPHQL =
  "# This file was automatically generated (by @graphitation/cli) and should not be edited.\n";

export function graphitation(): Command {
  const extractSchemaCommand = new Command();
  extractSchemaCommand
    .name("extract-schema")
    .argument("<files...>")
    .description(
      "extract implicit resolvers to a ts file from graphql typedefs",
    )
    .action(async (files: Array<string>) => {
      await typeDefsToImplicitResolversImpl(await getFiles(files));
    });

  const generateInterfacesCommand = new Command();
  generateInterfacesCommand
    .name("generate-interfaces")
    .argument("<inputs...>")
    .option(
      "-o, --output-dir [outputDir]",
      "output directory relative to file, default generated",
    )
    .option(
      "-ci, --context-import [contextImport]",
      "from where to import context",
    )
    .option("-cn, --context-name [contextName]", "Context name")
    .option("-ei, --enums-import [enumsImport]", "from where to import enums")
    .option("-l, --legacy", "generate legacy types")
    .option("--legacy-models", "do not use models for object types")
    .option("--scope [scope]", "generate models only for scope")
    .description("generate interfaces and models")
    .action(
      async (inputs: Array<string>, options: GenerateInterfacesOptions) => {
        await generateInterfaces(await getFiles(inputs), options);
      },
    );

  const mergeSchemasCommand = new Command();
  mergeSchemasCommand
    .name("merge-schemas")
    .argument("<entryPoints...>")
    .option("-o,--output [output]", "output file, otherwise writes to stdout")
    .option(
      "--ignore-entry-point-missing-types",
      "ignore missing types in entry points for implicit types compat mode",
    )
    .action(
      async (entryPoints: Array<string>, options: MergeSchemasOptions) => {
        await mergeSchemas(await getFiles(entryPoints), options);
      },
    );

  return program
    .name("supermassive")
    .addCommand(extractSchemaCommand)
    .addCommand(generateInterfacesCommand)
    .addCommand(mergeSchemasCommand);
}

async function getFiles(inputs: Array<string>): Promise<Array<string>> {
  return Promise.all(
    inputs
      .map((input) => {
        if (fsSync.existsSync(input)) {
          return input;
        } else {
          return glob.sync([input]);
        }
      })
      .flat()
      .filter(Boolean)
      .map(async (file) => {
        let fullPath: string;
        if (path.isAbsolute(file)) {
          fullPath = file;
        } else {
          fullPath = path.join(process.cwd(), file);
        }
        const stat = await fs.stat(fullPath);
        if (!stat.isFile) {
          throw new Error(`Invalid file ${file}`);
        }
        return fullPath;
      }),
  );
}
function getContextPath(outputDir: string, contextImport: string | undefined) {
  if (!contextImport) {
    return;
  }

  if (!contextImport.startsWith(".")) {
    return contextImport;
  }

  const contextDir = path.join(process.cwd(), contextImport);

  return path
    .relative(outputDir, contextDir)
    .split(path.sep)
    .join(path.posix.sep);
}

async function generateInterfaces(
  files: Array<string>,
  options: GenerateInterfacesOptions,
): Promise<void> {
  for (const file of files) {
    const content = await fs.readFile(file, { encoding: "utf-8" });
    const document = parse(content);

    const outputPath = path.join(
      path.dirname(file),
      options.outputDir ? options.outputDir : "__generated__",
    );

    const result = generateTS(document, {
      outputPath,
      documentPath: file,
      contextImport: getContextPath(outputPath, options.contextImport) || null,
      contextName: options.contextName,
      enumsImport: getContextPath(outputPath, options.enumsImport) || null,
      legacyCompat: !!options.legacy,
      legacyNoModelsForObjects: !!options.legacyModels,
      modelScope: options.scope || null,
    });

    await fs.mkdir(outputPath, { recursive: true });

    const printer = ts.createPrinter();
    const outputs = result.files.map((file) =>
      fs.writeFile(
        path.join(outputPath, file.fileName),
        PREPEND_TO_INTERFACES +
          printer.printNode(ts.EmitHint.SourceFile, file, file),
        { encoding: "utf-8" },
      ),
    );

    await Promise.all(outputs);
  }
}

async function typeDefsToImplicitResolversImpl(
  files: Array<string>,
): Promise<void> {
  for (const file of files) {
    const content = await fs.readFile(file, { encoding: "utf-8" });
    const document = parse(content);

    const tsContents = extractImplicitTypesToTypescript(document);
    const tsDir = path.join(path.dirname(file), "__generated__");
    await fs.mkdir(tsDir, { recursive: true });
    const tsFileName = path.join(
      tsDir,
      path.basename(file, path.extname(file)) + ".ts",
    );
    const printer = ts.createPrinter();

    await fs.writeFile(
      tsFileName,
      printer.printNode(ts.EmitHint.SourceFile, tsContents, tsContents),
      { encoding: "utf-8" },
    );
  }
}

async function mergeSchemas(
  files: Array<string>,
  options: MergeSchemasOptions,
): Promise<void> {
  const result = await mergeSchemasImpl(
    files.map((absolutePath) => ({ absolutePath })),
    new FileSystemModuleLoader(),
  );
  let fail = false;
  if (result.errors) {
    for (const error of result.errors || []) {
      if (!(options.ignoreEntryPointMissingTypes && error.isEntryPoint)) {
        fail = true;
      }
      console.warn(printMergeSchemaError(error));
    }
  }
  if (fail) {
    throw new Error("Missing types, aborting.");
  } else {
    const output = PREPEND_TO_GRAPHQL + print(result.document);
    if (options.output) {
      await fs.writeFile(options.output, output, { encoding: "utf-8" });
    } else {
      process.stdout.write(output);
    }
  }
}
