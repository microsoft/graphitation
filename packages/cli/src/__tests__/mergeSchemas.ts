import fs from "fs/promises";
import path from "path";
import { graphitation } from "../graphitation";

describe(graphitation, () => {
  beforeEach(async () => {
    await fs.rm(
      path.join(__dirname, "fixtures", "mergeSchemas", "output.graphql"),
      {
        recursive: true,
        force: true,
      },
    );
  });

  afterEach(async () => {
    jest.restoreAllMocks().clearAllMocks();
    await fs.rm(
      path.join(__dirname, "fixtures", "mergeSchemas", "output.graphql"),
      {
        recursive: true,
        force: true,
      },
    );
  });

  it("generates from single file", async () => {
    const program = graphitation();
    const output = path.join(
      __dirname,
      "./fixtures/mergeSchemas/output.graphql",
    );
    await program.parseAsync([
      "node",
      "graphitation",
      "merge-schemas",
      "-o",
      output,
      path.join(__dirname, "./fixtures/mergeSchemas/entry-point.graphql"),
    ]);
    expect(await fs.readFile(output, { encoding: "utf-8" })).toMatchSnapshot();
  });

  it("generates from glob", async () => {
    const program = graphitation();
    const output = path.join(
      __dirname,
      "./fixtures/mergeSchemas/output.graphql",
    );
    await program.parseAsync([
      "node",
      "graphitation",
      "merge-schemas",
      "-o",
      output,
      path.join(__dirname, "./fixtures/mergeSchemas/test-*.graphql"),
    ]);
    expect(await fs.readFile(output, { encoding: "utf-8" })).toMatchSnapshot();
  });

  it("complains about missing imports", async () => {
    const warn = jest.spyOn(console, "warn");
    const program = graphitation();
    const output = path.join(
      __dirname,
      "./fixtures/mergeSchemas/output.graphql",
    );
    await expect(
      program.parseAsync([
        "node",
        "graphitation",
        "merge-schemas",
        "-o",
        output,
        path.join(__dirname, "./fixtures/mergeSchemas/missing-types.graphql"),
      ]),
    ).rejects.toEqual(expect.any(Error));
    // Extended type, DateTime and wrong import
    expect(warn).toHaveBeenCalledTimes(3);
  });

  it("complains about missing imports, but allows passing with a flag", async () => {
    const warn = jest.spyOn(console, "warn");
    const program = graphitation();
    const output = path.join(
      __dirname,
      "./fixtures/mergeSchemas/output.graphql",
    );
    await program.parseAsync([
      "node",
      "graphitation",
      "merge-schemas",
      "--ignore-entry-point-missing-types",
      "-o",
      output,
      path.join(
        __dirname,
        "./fixtures/mergeSchemas/missing-types-entry-point.graphql",
      ),
    ]);

    expect(await fs.readFile(output, { encoding: "utf-8" })).toMatchSnapshot();
    // Extended type and DateTime
    expect(warn).toHaveBeenCalledTimes(2);
  });
});
