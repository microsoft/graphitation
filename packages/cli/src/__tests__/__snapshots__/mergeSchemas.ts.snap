// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphitation complains about missing imports, but allows passing with a flag 1`] = `
"# This file was automatically generated (by @graphitation/cli) and should not be edited.
extend type NonType {
  foo: DateTime
}
"
`;

exports[`graphitation generates from glob 1`] = `
"# This file was automatically generated (by @graphitation/cli) and should not be edited.
type Person implements Node {
  id: ID!
}

"""ISO8601 Date Time"""
scalar DateTime

"""
Define a type to be compliant with the Global Object Identification best-practice spec.
https://graphql.org/learn/global-object-identification/
"""
interface Node {
  id: ID!
}

type Bar implements Node {
  id: ID!
}

type Foo implements Node {
  id: ID!
  timestamp: DateTime
}

type Frobnar implements Node {
  id: ID!
}

extend type Person {
  foo: Foo
}
"
`;

exports[`graphitation generates from single file 1`] = `
"# This file was automatically generated (by @graphitation/cli) and should not be edited.
type Baz implements Node {
  id: ID!
  foo: Foo
}

type Person implements Node {
  id: ID!
}

"""ISO8601 Date Time"""
scalar DateTime

"""
Define a type to be compliant with the Global Object Identification best-practice spec.
https://graphql.org/learn/global-object-identification/
"""
interface Node {
  id: ID!
}

type Bar implements Node {
  id: ID!
}

type Foo implements Node {
  id: ID!
  timestamp: DateTime
}

extend type Person {
  foo: Foo
}
"
`;
